addons/l10n_vn_edi_viettel/models/account_move.py

```diff
@@ -0,0 +1,766 @@
+        string='Secret Code',
+        help='Secret code that can be used by a customer to lookup an invoice on SInvoice.',
+        copy=False,
+        readonly=True,
+    )
+    l10n_vn_edi_issue_date = fields.Datetime(
+        string='Issue Date',
+        help='Date of issue of the invoice on the e-invoicing system.',
+        copy=False,
+        readonly=True,
+    )
+    l10n_vn_edi_sinvoice_file_id = fields.Many2one(
+        comodel_name='ir.attachment',
+        compute=lambda self: self._compute_linked_attachment_id('l10n_vn_edi_sinvoice_file_id', 'l10n_vn_edi_sinvoice_file'),
+        depends=['l10n_vn_edi_sinvoice_file'],
+        export_string_translation=False,
+    )
+    l10n_vn_edi_sinvoice_file = fields.Binary(
+        string='SInvoice json File',
+        copy=False,
+        export_string_translation=False,
+    )
+    l10n_vn_edi_sinvoice_xml_file_id = fields.Many2one(
+        comodel_name='ir.attachment',
+        compute=lambda self: self._compute_linked_attachment_id('l10n_vn_edi_sinvoice_xml_file_id', 'l10n_vn_edi_sinvoice_xml_file'),
+        depends=['l10n_vn_edi_sinvoice_xml_file'],
+        export_string_translation=False,
+    )
+    l10n_vn_edi_sinvoice_xml_file = fields.Binary(
+        string='SInvoice xml File',
+        copy=False,
+        export_string_translation=False,
+    )
+    l10n_vn_edi_sinvoice_pdf_file_id = fields.Many2one(
+        comodel_name='ir.attachment',
+        compute=lambda self: self._compute_linked_attachment_id('l10n_vn_edi_sinvoice_pdf_file_id', 'l10n_vn_edi_sinvoice_pdf_file'),
+        depends=['l10n_vn_edi_sinvoice_pdf_file'],
+        export_string_translation=False,
+    )
+    l10n_vn_edi_sinvoice_pdf_file = fields.Binary(
+        string='SInvoice pdf File',
+        copy=False,
+        export_string_translation=False,
+    )
+    # Replacement/Adjustment fields
+    l10n_vn_edi_agreement_document_name = fields.Char(
+        string='Agreement Name',
+        copy=False,
+    )
+    l10n_vn_edi_agreement_document_date = fields.Datetime(
+        string='Agreement Date',
+        copy=False,
+    )
+    l10n_vn_edi_adjustment_type = fields.Selection(
+        string='Adjustment type',
+        selection=[
+            ('1', 'Money adjustment'),
+            ('2', 'Information adjustment'),
+        ],
+        copy=False,
+    )
+    # Only used in case of replacement invoice.
+    l10n_vn_edi_replacement_origin = fields.Many2one(
+        comodel_name='account.move',
+        string='Replacement of',
+        copy=False,
+        readonly=True,
+        check_company=True,
+        export_string_translation=False,
+    )
+
+    @api.depends('l10n_vn_edi_invoice_state')
+    def _compute_show_reset_to_draft_button(self):
+        # EXTEND 'account'
+        super()._compute_show_reset_to_draft_button()
+        self.filtered(lambda m: m._l10n_vn_edi_is_sent() and m.l10n_vn_edi_invoice_state != 'canceled').show_reset_to_draft_button = False
```

Comment:
It's a bit confusing. `_l10n_vn_edi_is_sent` contains a lot of statuses including `canceled`... Are you sure quite all the statuses need to be everywhere like that? It make the flow difficult to understand since it seems all actions can be called from every states :thinking:

Reply:
I mean, the point is that `_l10n_vn_edi_is_sent` returns true if the invoice exists on Viettel's systems.

- You never want to allow reset to draft as long as the invoice is in a "used" state (basically, if it's on the system but not cancelled)
- It would not make sense to allow trying to fetch the files data if you never sent the invoice
- If it is sent, you need to go through a cancel request. (although here I think we can also not include the already canceled ones)
- Last use is to validate that a out_refund that you would want to send is linked to an invoice that was actually sent, so it makes sense to check all these states.

I understand it may not be super clear, I'm open to suggestions. But at least this way, you can understand that "this action should only/never be done when the invoice has been set"